pyomo/contrib/doe/measurements.py

```diff
@@ -25,252 +25,273 @@
+                    all_index_list.append(index_list)
+            if time_index:
+                all_index_list.append(time_index[i])
+
+            # all idnex list for one variable, such as ["CA", 10, 1]
+            all_index_for_var = list(itertools.product(*all_index_list))
+
+            for lst in all_index_for_var:
+                name1 = name_data+"["
+                for i, idx in enumerate(lst):
+                    name1 += str(idx)
+
+                    # if i is the last index, close the []. if not, add a "," for the next index. 
+                    if i==len(lst)-1:
+                        name1 += "]"
+                    else:
+                        name1 += ","
+
+                self.special_set.append(name1)
+    
+class Measurements(SpecialSet):
+    def __init__(self):
+        """
+        This class stores information on which algebraic and differential variables in the Pyomo model are considered measurements. 
+        """
+        super().__init__()
+
+    def specify(self, self_define_res):
         """
-        flatten_names = []
-        for j in measure_name_and_index.keys():
-            if measure_name_and_index[j] is not None: # if it has extra index
-                for ind in measure_name_and_index[j]:
-                    flatten_name = j + self.ind_string + str(ind)
-                    flatten_names.append(flatten_name)
-            else:
-                flatten_names.append(j)
-
-        self.flatten_measure_name = flatten_names
-
-    def _generate_variance(self, flatten_measure_name, variance, name_and_index):
+        User can pass already defined measurement names here
+        """
+        self.measurement_name = super().specify(self_define_res)
+
+        # generate default variance
+        self._generate_variance()
+
+    def add_elements(self, var_name, extra_index=None, time_index=[0]):
         """
-        Generate the variance dictionary
         Parameters
+        -----------
+        var_name: a ``list`` of measurement var names 
+            extra_index: a ``list`` containing extra indexes except for time indexes 
+                if default (None), no extra indexes needed for all var in var_name
+                if it is a nested list, it is a ``list`` of ``list`` of ``list``, 
+                they are different multiple sets of indexes for different var_name
+                for e.g., extra_index[0] are all indexes for var_name[0], extra_index[0][0] are the first index for var_name[0]
+            time_index: a ``list`` containing time indexes
+                default choice is [0], means this is an algebraic variable
+                if it is a nested list, it is a ``list`` of ``lists``, they are different time set for different var in var_name
+        """
+        self._check_names(var_name, extra_index, time_index)
+        self.measurement_name =  super().add_elements(var_name=var_name, extra_index=extra_index, time_index=time_index)
```

Comment:
Why not just call this "names"? That would make it easier to generalize across Measurements and DesignVariables.